src/app/store/observeMiddleware.ts

```diff
@@ -0,0 +1,95 @@
+  id: string;
+  /**
+   * A custom quality function. It not provided Objest.is will be used.
+   */
+  equals?: (a: unknown, b: unknown) => boolean;
+  /**
+   * Function to create "something" which will be used to determine if the side effect should run.
+   * deepEqual is used for the comparison, so use primitives, shallow objects or arrays to keep
+   * performance reasonable.
+   */
+  getObserved: (rootState: RootState) => T;
+  /**
+   * Runs the side effect providing both the previous and current version of the derrived state.
+   */
+  sideEffect: (states: { previous: T; current: T }) => void;
+}
+
+function isObserveAction(action: unknown): action is ReturnType<typeof observe> {
+  return (
+    typeof action === 'object' &&
+    action !== null &&
+    'type' in action &&
+    action.type === 'observer/OBSERVE'
+  );
+}
+
+// Need to user a higher order function to get the correct typings and inference, it will now correctly
+// type the value of T based on what the getObserved returns
+export const observe = <T>(storeObserver: StoreObserver<T>) =>
+  createCustomAction('observer/OBSERVE', (storeObserver: StoreObserver<T>) => ({ storeObserver }))(
+    storeObserver,
+  );
+export const unobserve = createAction('observer/UNOBSERVE')<string>();
+export const clearObservers = createAction('observer/CLEAR_OBSERVERS')();
+
+export function observerMiddleware<D extends Dispatch>(
+  api: MiddlewareAPI<D, RootState>,
+): ReturnType<Middleware> {
+  const observers = new Map<string, StoreObserver<unknown>>();
+  return (next) => (action) => {
+    // Taken from the redux listener middleware, apparently some actions may not
+    // be action objects, https://github.com/reduxjs/redux-toolkit/blob/0d1f7101e83865714cb512c850bc53ffaee2d5e5/packages/toolkit/src/listenerMiddleware/index.ts#L438-L440
+    if (!isAction(action)) {
+      return next(action);
+    }
+
+    if (isObserveAction(action)) {
+      observers.set(action.storeObserver.id, action.storeObserver);
+      return;
+    }
+
+    if (isActionOf(unobserve, action)) {
+      observers.delete(action.payload);
+      return;
+    }
+
+    if (isActionOf(clearObservers, action)) {
+      observers.clear();
+      return;
+    }
+
+    const previousRootState = api.getState();
+    // Forward the state model so that we can compare and look for changes.
+    const result = next(action);
+    const currentRootState = api.getState();
+
+    for (const [_id, observer] of observers) {
+      // Grab the slice or derivation the given observer cares about.
+      const previous = observer.getObserved(previousRootState);
+      const current = observer.getObserved(currentRootState);
+      const equals = observer.equals || Object.is;
+      if (!equals(previous, current)) {
+        if ($DIM_FLAVOR !== 'release') {
+          infoLog('observer sideEffect', `Running side effect for observer`, {
+            id: observer.id,
+            previous,
+            current,
+          });
+        }
```

Comment:
Added this as I found it useful for debugging but happy to remove based on how useful you think it might be.